### PR TITLE
Add support for generating conda binary packages for ROBOTOLOGY_USES_PYTHON set to ON

### PR DESCRIPTION
**Pull request by traversaro**
Fix https://github.com/robotology/robotology-superbuild/issues/641 . It a nutshell this boils down to stop setting `-DROBOTOLOGY_USES_PYTHON:BOOL=OFF` in the `generate-conda-packages` Action, and add the appropriate options to ensure that on Windows the Python bindings are installed in the right place when generating a conda package.

Furthermore, we add some logic to ensure that the correct run dependency of numpy is used, see https://conda-forge.org/docs/maintainer/knowledge_base.html#building-against-numpy .

### CODE REVIEW

**Comment by traversaro** [comment]
A first test run of the conde package generation logic is in progress in https://github.com/robotology/robotology-superbuild/runs/2631285243 . In the meanwhile, the PR is ready for review.

**Comment by traversaro** [comment]
New run of the `generate-conda-packages` GHA workflow after the last modifications:
* https://github.com/robotology/robotology-superbuild/actions/runs/861197279

File: cmake/BuildYARP.cmake [hunk 1]
```diff
@@ -38,6 +37,15 @@ else()
   set(ENABLE_yarpcar_mjpeg ON)
 endif()
 
+# For what regards Python installation, the options changes depending
+# on whater we are installing YARP from source, or we are generating a
```
Comment (nunoguedelha) [inline cmake/BuildYARP.cmake:41]:
```suggestion
# on whether we are installing YARP from source, or generating a
```

File: cmake/BuildiDynTree.cmake [hunk 2]
```diff
@@ -14,6 +14,17 @@ list(APPEND iDynTree_DEPENDS YARP)
 list(APPEND iDynTree_DEPENDS ICUB)
 list(APPEND iDynTree_DEPENDS OsqpEigen)
 
+# For what regards Python installation, the options changes depending
+# on whater we are installing YARP from source, or we are generating a
```
Comment (nunoguedelha) [inline cmake/BuildiDynTree.cmake:18]:
```suggestion
# on whether we are installing YARP from source, or generating a
```

**Review by nunoguedelha — CHANGES_REQUESTED** [review]
Just a couple of typos.

I'm not so familiar with all these code details but it seems ok.

**Comment by traversaro** [comment]
Thanks @nunoguedelha, I fixed the comment.

**Comment by traversaro** [comment]
The actual

> 
> 
> New run of the `generate-conda-packages` GHA workflow after the last modifications:
> 
>     * https://github.com/robotology/robotology-superbuild/actions/runs/861197279

New job after some fixes: https://github.com/robotology/robotology-superbuild/actions/runs/861887899 . The generation works fine, the generation time has reached 4h/4.30h due to the package that needs to be built multiple times for different version of Python, but for now the time is far from the 6h limit of GitHub Actions, and if necessary we can investigate how to decrease it.

**Review by nunoguedelha — APPROVED** [review]

**Comment by traversaro** [comment]
There was some problem in this part of the recipe template on macOS: 
~~~
  run:
{% if add_numpy_runtime_dep %}{% raw %}    - {{ pin_compatible('numpy') }}{% endraw %}{% endif %}
    - __osx >={{ MACOSX_DEPLOYMENT_TARGET|default("10.9") }}  # [osx and x86_64]
~~~

That resulted in this `conda build`  error:
~~~
2021-05-21T06:54:00.5001870Z /Users/runner/work/robotology-superbuild/robotology-superbuild/build/conda/generated_recipes/funny-things
2021-05-21T06:54:00.5003220Z -- CONDA_GENERATION_SCRIPT_RETURN_VALUE: 0
2021-05-21T06:54:00.5004640Z -- conda recipes correctly generated in /Users/runner/work/robotology-superbuild/robotology-superbuild/build/conda/generated_recipes.
2021-05-21T06:54:00.5006180Z -- To build the generated conda recipes, navigate to the directory and run conda build . in it.
2021-05-21T06:54:00.5007130Z -- Configuring done
2021-05-21T06:54:00.5007840Z -- Generating done
2021-05-21T06:54:00.5009020Z -- Build files have been written to: /Users/runner/work/robotology-superbuild/robotology-superbuild/build
2021-05-21T06:54:00.5061740Z ##[group]Run cd build/conda/generated_recipes
2021-05-21T06:54:00.5062460Z [36;1mcd build/conda/generated_recipes[0m
2021-05-21T06:54:00.5063140Z [36;1m# We use the conda_build_config from conda-forge-pinning, and our local one[0m
2021-05-21T06:54:00.5064520Z [36;1m# See https://docs.conda.io/projects/conda-build/en/latest/resources/variants.html#creating-conda-build-variant-config-files[0m
2021-05-21T06:54:00.5065940Z [36;1mconda build -m ${CONDA_PREFIX}/conda_build_config.yaml -m ${GITHUB_WORKSPACE}/conda/conda_build_config.yml .[0m
2021-05-21T06:54:00.5233070Z shell: /bin/bash -l {0}
2021-05-21T06:54:00.5233420Z env:
2021-05-21T06:54:00.5233850Z   CONDA_PKGS_DIR: /Users/runner/conda_pkgs_dir
2021-05-21T06:54:00.5234710Z   GHA_Matlab_ROOT_DIR: /Users/runner/work/robotology-superbuild/robotology-superbuild/msdk_R2020a_mexmaci64
2021-05-21T06:54:00.5235540Z   GHA_Matlab_MEX_EXTENSION: mexmaci64
2021-05-21T06:54:00.5235960Z ##[endgroup]
2021-05-21T06:54:03.7828800Z No numpy version specified in conda_build_config.yaml.  Falling back to default numpy value of 1.16
2021-05-21T06:54:03.7877790Z WARNING:conda_build.metadata:No numpy version specified in conda_build_config.yaml.  Falling back to default numpy value of 1.16
2021-05-21T06:54:03.8855110Z INFO:conda_build.variants:Adding in variants from internal_defaults
2021-05-21T06:54:03.8855890Z Adding in variants from internal_defaults
2021-05-21T06:54:03.8941360Z INFO:conda_build.variants:Adding in variants from /usr/local/miniconda/envs/test/conda_build_config.yaml
2021-05-21T06:54:03.8942440Z Adding in variants from /usr/local/miniconda/envs/test/conda_build_config.yaml
2021-05-21T06:54:03.9012270Z INFO:conda_build.variants:Adding in variants from /Users/runner/work/robotology-superbuild/robotology-superbuild/conda/conda_build_config.yml
2021-05-21T06:54:03.9014220Z Adding in variants from /Users/runner/work/robotology-superbuild/robotology-superbuild/conda/conda_build_config.yml
2021-05-21T06:54:04.7335260Z INFO:conda_build.metadata:Attempting to finalize metadata for bipedal-locomotion-framework
2021-05-21T06:54:04.7336900Z Attempting to finalize metadata for bipedal-locomotion-framework
2021-05-21T06:55:28.6166640Z Collecting package metadata (repodata.json): ...working... done
2021-05-21T06:55:28.6315790Z Solving environment: ...working... failed
2021-05-21T06:55:28.8353460Z INFO:conda_build.variants:Adding in variants from config.variant
2021-05-21T06:55:28.8354330Z Adding in variants from config.variant
2021-05-21T06:55:28.8355650Z INFO conda_build.variants:_combine_spec_dictionaries(234): Adding in variants from config.variant
2021-05-21T06:55:30.1107950Z Missing dependency yarp, but found recipe directory, so building yarp first
2021-05-21T06:55:30.1111570Z Missing dependency lie-group-controllers, but found recipe directory, so building lie-group-controllers first
2021-05-21T06:55:30.1113560Z Missing dependency matio-cpp, but found recipe directory, so building matio-cpp first
2021-05-21T06:55:42.1124190Z Collecting package metadata (repodata.json): ...working... done
2021-05-21T06:55:42.1268910Z Solving environment: ...working... failed
2021-05-21T06:55:42.1954020Z Traceback (most recent call last):
2021-05-21T06:55:42.1956780Z   File "/usr/local/miniconda/envs/test/lib/python3.8/site-packages/conda_build/environ.py", line 802, in get_install_actions
2021-05-21T06:55:42.1958730Z     actions = install_actions(prefix, index, specs, force=True)
2021-05-21T06:55:42.1960550Z   File "/usr/local/miniconda/envs/test/lib/python3.8/site-packages/conda/common/io.py", line 88, in decorated
2021-05-21T06:55:42.1961650Z     return f(*args, **kwds)
2021-05-21T06:55:42.1963100Z   File "/usr/local/miniconda/envs/test/lib/python3.8/site-packages/conda/plan.py", line 474, in install_actions
2021-05-21T06:55:42.1964400Z     txn = solver.solve_for_transaction(prune=prune, ignore_pinned=not pinned)
2021-05-21T06:55:42.1969630Z   File "/usr/local/miniconda/envs/test/lib/python3.8/site-packages/conda/core/solve.py", line 114, in solve_for_transaction
2021-05-21T06:55:42.1970860Z     unlink_precs, link_precs = self.solve_for_diff(update_modifier, deps_modifier,
2021-05-21T06:55:42.1972370Z   File "/usr/local/miniconda/envs/test/lib/python3.8/site-packages/conda/core/solve.py", line 157, in solve_for_diff
2021-05-21T06:55:42.1973420Z     final_precs = self.solve_final_state(update_modifier, deps_modifier, prune, ignore_pinned,
2021-05-21T06:55:42.1975000Z   File "/usr/local/miniconda/envs/test/lib/python3.8/site-packages/conda/core/solve.py", line 275, in solve_final_state
2021-05-21T06:55:42.1975930Z     ssc = self._add_specs(ssc)
2021-05-21T06:55:42.1977150Z   File "/usr/local/miniconda/envs/test/lib/python3.8/site-packages/conda/core/solve.py", line 565, in _add_specs
2021-05-21T06:55:42.1978010Z     explicit_pool = ssc.r._get_package_pool(self.specs_to_add)
2021-05-21T06:55:42.1979330Z   File "/usr/local/miniconda/envs/test/lib/python3.8/site-packages/conda/resolve.py", line 555, in _get_package_pool
2021-05-21T06:55:42.1980170Z     pool = self.get_reduced_index(specs)
2021-05-21T06:55:42.1981470Z   File "/usr/local/miniconda/envs/test/lib/python3.8/site-packages/conda/common/io.py", line 88, in decorated
2021-05-21T06:55:42.1982300Z     return f(*args, **kwds)
2021-05-21T06:55:42.1983550Z   File "/usr/local/miniconda/envs/test/lib/python3.8/site-packages/conda/resolve.py", line 576, in get_reduced_index
2021-05-21T06:55:42.1985170Z     explicit_specs, features = self.verify_specs(explicit_specs)
2021-05-21T06:55:42.1986600Z   File "/usr/local/miniconda/envs/test/lib/python3.8/site-packages/conda/resolve.py", line 288, in verify_specs
2021-05-21T06:55:42.1987770Z     raise ResolvePackageNotFound(bad_deps)
2021-05-21T06:55:42.1988730Z conda.exceptions.ResolvePackageNotFound: 
2021-05-21T06:55:42.1989810Z   - yarp
2021-05-21T06:55:42.1990660Z   - lie-group-controllers
2021-05-21T06:55:42.1991470Z   - matio-cpp
2021-05-21T06:55:42.1992850Z   - idyntree
2021-05-21T06:55:42.1993150Z 
2021-05-21T06:55:42.1993720Z During handling of the above exception, another exception occurred:
2021-05-21T06:55:42.1994210Z 
2021-05-21T06:55:42.1994650Z Traceback (most recent call last):
2021-05-21T06:55:42.1995860Z   File "/usr/local/miniconda/envs/test/lib/python3.8/site-packages/conda_build/build.py", line 3068, in build_tree
2021-05-21T06:55:42.1996830Z     packages_from_this = build(metadata, stats,
2021-05-21T06:55:42.1998110Z   File "/usr/local/miniconda/envs/test/lib/python3.8/site-packages/conda_build/build.py", line 2031, in build
2021-05-21T06:55:42.1999020Z     output_metas = expand_outputs([(m, need_source_download, need_reparse_in_env)])
2021-05-21T06:55:42.2000450Z   File "/usr/local/miniconda/envs/test/lib/python3.8/site-packages/conda_build/render.py", line 789, in expand_outputs
2021-05-21T06:55:42.2001650Z     for (output_dict, m) in deepcopy(_m).get_output_metadata_set(permit_unsatisfiable_variants=False):
2021-05-21T06:55:42.2003280Z   File "/usr/local/miniconda/envs/test/lib/python3.8/site-packages/conda_build/metadata.py", line 2115, in get_output_metadata_set
2021-05-21T06:55:42.2004330Z     conda_packages = finalize_outputs_pass(ref_metadata, conda_packages, pass_no=0,
2021-05-21T06:55:42.2005830Z   File "/usr/local/miniconda/envs/test/lib/python3.8/site-packages/conda_build/metadata.py", line 771, in finalize_outputs_pass
2021-05-21T06:55:42.2006830Z     parent_metadata.parse_until_resolved()
2021-05-21T06:55:42.2008270Z   File "/usr/local/miniconda/envs/test/lib/python3.8/site-packages/conda_build/metadata.py", line 1079, in parse_until_resolved
2021-05-21T06:55:42.2020360Z     self.parse_again(permit_undefined_jinja=False,
2021-05-21T06:55:42.2022160Z   File "/usr/local/miniconda/envs/test/lib/python3.8/site-packages/conda_build/metadata.py", line 1001, in parse_again
2021-05-21T06:55:42.2023110Z     self.meta = parse(self._get_contents(permit_undefined_jinja,
2021-05-21T06:55:42.2024480Z   File "/usr/local/miniconda/envs/test/lib/python3.8/site-packages/conda_build/metadata.py", line 1596, in _get_contents
2021-05-21T06:55:42.2025990Z     rendered = template.render(environment=env)
2021-05-21T06:55:42.2027420Z   File "/usr/local/miniconda/envs/test/lib/python3.8/site-packages/jinja2/environment.py", line 1304, in render
2021-05-21T06:55:42.2028300Z     self.environment.handle_exception()
2021-05-21T06:55:42.2029790Z   File "/usr/local/miniconda/envs/test/lib/python3.8/site-packages/jinja2/environment.py", line 925, in handle_exception
2021-05-21T06:55:42.2030680Z     raise rewrite_traceback_stack(source=source)
2021-05-21T06:55:42.2032500Z   File "/Users/runner/work/robotology-superbuild/robotology-superbuild/build/conda/generated_recipes/bipedal-locomotion-framework/meta.yaml", line 49, in top-level template code
2021-05-21T06:55:42.2033780Z     
2021-05-21T06:55:42.2034940Z   File "/usr/local/miniconda/envs/test/lib/python3.8/site-packages/conda_build/jinja_context.py", line 230, in pin_compatible
2021-05-21T06:55:42.2036300Z     pins, _, _ = get_env_dependencies(m, 'host', m.config.variant)
2021-05-21T06:55:42.2037720Z   File "/usr/local/miniconda/envs/test/lib/python3.8/site-packages/conda_build/render.py", line 139, in get_env_dependencies
2021-05-21T06:55:42.2038710Z     actions = environ.get_install_actions(tmpdir, tuple(dependencies), env,
2021-05-21T06:55:42.2040140Z   File "/usr/local/miniconda/envs/test/lib/python3.8/site-packages/conda_build/environ.py", line 804, in get_install_actions
2021-05-21T06:55:42.2041630Z     raise DependencyNeedsBuildingError(exc, subdir=subdir)
2021-05-21T06:55:42.2043860Z conda_build.exceptions.DependencyNeedsBuildingError: Unsatisfiable dependencies for platform osx-64: {'matio-cpp', 'yarp', 'lie-group-controllers', 'idyntree'}
2021-05-21T06:55:42.2045090Z 
2021-05-21T06:55:42.2045770Z During handling of the above exception, another exception occurred:
2021-05-21T06:55:42.2046240Z 
2021-05-21T06:55:42.2046630Z Traceback (most recent call last):
2021-05-21T06:55:42.2047930Z   File "/usr/local/miniconda/envs/test/lib/python3.8/site-packages/conda_build/environ.py", line 802, in get_install_actions
2021-05-21T06:55:42.2049020Z     actions = install_actions(prefix, index, specs, force=True)
2021-05-21T06:55:42.2050300Z   File "/usr/local/miniconda/envs/test/lib/python3.8/site-packages/conda/common/io.py", line 88, in decorated
2021-05-21T06:55:42.2051060Z     return f(*args, **kwds)
2021-05-21T06:55:42.2052240Z   File "/usr/local/miniconda/envs/test/lib/python3.8/site-packages/conda/plan.py", line 474, in install_actions
2021-05-21T06:55:42.2053160Z     txn = solver.solve_for_transaction(prune=prune, ignore_pinned=not pinned)
2021-05-21T06:55:42.2054580Z   File "/usr/local/miniconda/envs/test/lib/python3.8/site-packages/conda/core/solve.py", line 114, in solve_for_transaction
2021-05-21T06:55:42.2055660Z     unlink_precs, link_precs = self.solve_for_diff(update_modifier, deps_modifier,
2021-05-21T06:55:42.2057130Z   File "/usr/local/miniconda/envs/test/lib/python3.8/site-packages/conda/core/solve.py", line 157, in solve_for_diff
2021-05-21T06:55:42.2058160Z     final_precs = self.solve_final_state(update_modifier, deps_modifier, prune, ignore_pinned,
2021-05-21T06:55:42.2059620Z   File "/usr/local/miniconda/envs/test/lib/python3.8/site-packages/conda/core/solve.py", line 275, in solve_final_state
2021-05-21T06:55:42.2060370Z     ssc = self._add_specs(ssc)
2021-05-21T06:55:42.2061550Z   File "/usr/local/miniconda/envs/test/lib/python3.8/site-packages/conda/core/solve.py", line 565, in _add_specs
2021-05-21T06:55:42.2062390Z     explicit_pool = ssc.r._get_package_pool(self.specs_to_add)
2021-05-21T06:55:42.2063710Z   File "/usr/local/miniconda/envs/test/lib/python3.8/site-packages/conda/resolve.py", line 555, in _get_package_pool
2021-05-21T06:55:42.2064670Z     pool = self.get_reduced_index(specs)
2021-05-21T06:55:42.2065870Z   File "/usr/local/miniconda/envs/test/lib/python3.8/site-packages/conda/common/io.py", line 88, in decorated
2021-05-21T06:55:42.2066630Z     return f(*args, **kwds)
2021-05-21T06:55:42.2067820Z   File "/usr/local/miniconda/envs/test/lib/python3.8/site-packages/conda/resolve.py", line 576, in get_reduced_index
2021-05-21T06:55:42.2069280Z     explicit_specs, features = self.verify_specs(explicit_specs)
2021-05-21T06:55:42.2070760Z   File "/usr/local/miniconda/envs/test/lib/python3.8/site-packages/conda/resolve.py", line 288, in verify_specs
2021-05-21T06:55:42.2071680Z     raise ResolvePackageNotFound(bad_deps)
2021-05-21T06:55:42.2072570Z conda.exceptions.ResolvePackageNotFound: 
2021-05-21T06:55:42.2073660Z   - icub-main
2021-05-21T06:55:42.2074330Z   - yarp
2021-05-21T06:55:42.2075090Z   - osqp-eigen
2021-05-21T06:55:42.2075370Z 
2021-05-21T06:55:42.2075910Z During handling of the above exception, another exception occurred:
2021-05-21T06:55:42.2076380Z 
2021-05-21T06:55:42.2076890Z Traceback (most recent call last):
2021-05-21T06:55:42.2077930Z   File "/usr/local/miniconda/envs/test/bin/conda-build", line 11, in <module>
2021-05-21T06:55:42.2078560Z     sys.exit(main())
2021-05-21T06:55:42.2079730Z   File "/usr/local/miniconda/envs/test/lib/python3.8/site-packages/conda_build/cli/main_build.py", line 481, in main
2021-05-21T06:55:42.2080460Z     execute(sys.argv[1:])
2021-05-21T06:55:42.2081660Z   File "/usr/local/miniconda/envs/test/lib/python3.8/site-packages/conda_build/cli/main_build.py", line 470, in execute
2021-05-21T06:55:42.2082660Z     outputs = api.build(args.recipe, post=args.post, test_run_post=args.test_run_post,
2021-05-21T06:55:42.2084310Z   File "/usr/local/miniconda/envs/test/lib/python3.8/site-packages/conda_build/api.py", line 186, in build
2021-05-21T06:55:42.2085030Z     return build_tree(
2021-05-21T06:55:42.2086270Z   File "/usr/local/miniconda/envs/test/lib/python3.8/site-packages/conda_build/build.py", line 3187, in build_tree
2021-05-21T06:55:42.2087210Z     dep_metas = render_recipe(recipe_dir, config=metadata.config)
2021-05-21T06:55:42.2088580Z   File "/usr/local/miniconda/envs/test/lib/python3.8/site-packages/conda_build/render.py", line 861, in render_recipe
2021-05-21T06:55:42.2089460Z     rendered_metadata = distribute_variants(m, variants,
2021-05-21T06:55:42.2090820Z   File "/usr/local/miniconda/envs/test/lib/python3.8/site-packages/conda_build/render.py", line 764, in distribute_variants
2021-05-21T06:55:42.2091900Z     mv.parse_until_resolved(allow_no_other_outputs=allow_no_other_outputs,
2021-05-21T06:55:42.2093310Z   File "/usr/local/miniconda/envs/test/lib/python3.8/site-packages/conda_build/metadata.py", line 1079, in parse_until_resolved
2021-05-21T06:55:42.2094240Z     self.parse_again(permit_undefined_jinja=False,
2021-05-21T06:55:42.2095560Z   File "/usr/local/miniconda/envs/test/lib/python3.8/site-packages/conda_build/metadata.py", line 1001, in parse_again
2021-05-21T06:55:42.2096450Z     self.meta = parse(self._get_contents(permit_undefined_jinja,
2021-05-21T06:55:42.2097880Z   File "/usr/local/miniconda/envs/test/lib/python3.8/site-packages/conda_build/metadata.py", line 1596, in _get_contents
2021-05-21T06:55:42.2098770Z     rendered = template.render(environment=env)
2021-05-21T06:55:42.2100120Z   File "/usr/local/miniconda/envs/test/lib/python3.8/site-packages/jinja2/environment.py", line 1304, in render
2021-05-21T06:55:42.2101040Z     self.environment.handle_exception()
2021-05-21T06:55:42.2102370Z   File "/usr/local/miniconda/envs/test/lib/python3.8/site-packages/jinja2/environment.py", line 925, in handle_exception
2021-05-21T06:55:42.2103300Z     raise rewrite_traceback_stack(source=source)
2021-05-21T06:55:42.2104880Z   File "/Users/runner/work/robotology-superbuild/robotology-superbuild/build/conda/generated_recipes/idyntree/meta.yaml", line 47, in top-level template code
2021-05-21T06:55:42.2105830Z     
2021-05-21T06:55:42.2107000Z   File "/usr/local/miniconda/envs/test/lib/python3.8/site-packages/conda_build/jinja_context.py", line 230, in pin_compatible
2021-05-21T06:55:42.2108330Z     pins, _, _ = get_env_dependencies(m, 'host', m.config.variant)
2021-05-21T06:55:42.2109700Z   File "/usr/local/miniconda/envs/test/lib/python3.8/site-packages/conda_build/render.py", line 139, in get_env_dependencies
2021-05-21T06:55:42.2111070Z     actions = environ.get_install_actions(tmpdir, tuple(dependencies), env,
2021-05-21T06:55:42.2112600Z   File "/usr/local/miniconda/envs/test/lib/python3.8/site-packages/conda_build/environ.py", line 804, in get_install_actions
2021-05-21T06:55:42.2113620Z     raise DependencyNeedsBuildingError(exc, subdir=subdir)
2021-05-21T06:55:42.2115550Z conda_build.exceptions.DependencyNeedsBuildingError: Unsatisfiable dependencies for platform osx-64: {'icub-main', 'yarp', 'osqp-eigen'}
2021-05-21T06:55:45.3354850Z ##[error]Process completed with exit code 1.
2021-05-21T06:55:45.3511940Z Post job cleanup.
~~~

I temporary avoided the problem on macOS with this modification:
~~~
  run:
{% if add_numpy_runtime_dep %}{% raw %}    - {{ pin_compatible('numpy') }}{% endraw %}{% endif %}  # [not osx]
{% if add_numpy_runtime_dep %}    - numpy {% endif %}  # [osx]
    - __osx >={{ MACOSX_DEPLOYMENT_TARGET|default("10.9") }}  # [osx and x86_64]
~~~

This is not completly correct (it could give problem for iDynTree once Numpy 2 is released) but for the time being I think it is fine. I would open an issue to track improvments to that part that in any case has some other problems

**Comment by traversaro** [comment]
Even with the workarounds, the iDynTree build on macOS is failing  (https://github.com/robotology/robotology-superbuild/runs/2638342316?check_suite_focus=true) with error:
~~~
2021-05-21T12:09:58.5834710Z Verifying transaction: ...working... done
2021-05-21T12:10:01.9707380Z Executing transaction: ...working... done
2021-05-21T12:10:02.0698010Z Cloning into bare repository '/usr/local/miniconda/envs/test/conda-bld/git_cache/github.com/robotology/idyntree.git'...
2021-05-21T12:10:05.6592710Z Cloning into '/usr/local/miniconda/envs/test/conda-bld/idyntree_1621598792966/work'...
2021-05-21T12:10:05.6843020Z done.
2021-05-21T12:10:06.0837270Z HEAD is now at 0eb8ef3fa Merge pull request #869 from robotology/devel
2021-05-21T12:10:19.0515890Z CMake Error at /usr/local/miniconda/envs/test/conda-bld/idyntree_1621598792966/_build_env/share/cmake-3.20/Modules/FindPackageHandleStandardArgs.cmake:230 (message):
2021-05-21T12:10:19.0517780Z   Could NOT find Python3 (missing: Python3_NumPy_INCLUDE_DIRS NumPy) (found
2021-05-21T12:10:19.0518360Z   version "3.9.5")
2021-05-21T12:10:19.0518920Z Call Stack (most recent call first):
2021-05-21T12:10:19.0520570Z   /usr/local/miniconda/envs/test/conda-bld/idyntree_1621598792966/_build_env/share/cmake-3.20/Modules/FindPackageHandleStandardArgs.cmake:594 (_FPHSA_FAILURE_MESSAGE)
2021-05-21T12:10:19.0522520Z   /usr/local/miniconda/envs/test/conda-bld/idyntree_1621598792966/_build_env/share/cmake-3.20/Modules/FindPython/Support.cmake:3165 (find_package_handle_standard_args)
2021-05-21T12:10:19.0524140Z   /usr/local/miniconda/envs/test/conda-bld/idyntree_1621598792966/_build_env/share/cmake-3.20/Modules/FindPython3.cmake:485 (include)
2021-05-21T12:10:19.0524990Z   bindings/CMakeLists.txt:22 (find_package)
2021-05-21T12:10:19.0525340Z 
~~~

The issue seems to be that CMake is finding the Python3 (3.9) in the parent environment, rather that the Python in the build environment. This is is usually solved by adding `-DPython3_EXECUTABLE=%PYTHON%` (on Windows) or       `-DPython3_EXECUTABLE=$PYTHON` (on Linux/macOS) to ensure that the correct Python is selected (see https://github.com/robotology/yarp/issues/2509#issuecomment-785706289).

**Comment by traversaro** [comment]
> The issue seems to be that CMake is finding the Python3 (3.9) in the parent environment, rather that the Python in the build environment. This is is usually solved by adding `-DPython3_EXECUTABLE=%PYTHON%` (on Windows) or `-DPython3_EXECUTABLE=$PYTHON` (on Linux/macOS) to ensure that the correct Python is selected (see [robotology/yarp#2509 (comment)](https://github.com/robotology/yarp/issues/2509#issuecomment-785706289)).

This indeed fixed the issue. The two remaining open issues are:
* With standard `conda build .` the build time goes over the 6 hours of GitHub Actions on Linux and macOS. To avoid that, I switched to use `conda mambabuild`, but unfortunately that does not resolve automatically the build order of the dependency, so for the time being I expanded it manually myself in the GitHub Actions script.
* For some reason, while the different variants of Python packages for different version of Python are correctly build on Linux and macOS, on Windows only the version that uses Python 3.9 is produced. I am not sure why this is happening.

**Comment by traversaro** [comment]
> For some reason, while the different variants of Python packages for different version of Python are correctly build on Linux and macOS, on Windows only the version that uses Python 3.9 is produced. I am not sure why this is happening.

It turns out that this is a recipe generation problem. For some reason on Windows the `python` run dependency is not generated, not sure why (I am not even able to reproduce this locally).

**Comment by traversaro** [comment]
>     * With standard `conda build .` the build time goes over the 6 hours of GitHub Actions on Linux and macOS. To avoid that, I switched to use `conda mambabuild`, but unfortunately that does not resolve automatically the build order of the dependency, so for the time being I expanded it manually myself in the GitHub Actions script.

Related boa issue: https://github.com/mamba-org/boa/issues/117 .

**Comment by traversaro** [comment]
> For some reason, while the different variants of Python packages for different version of Python are correctly build on Linux and macOS, on Windows only the version that uses Python 3.9 is produced. I am not sure why this is happening.

I was definitely barking at the wrong tree here. The problem is simply that `ROBOTOLOGY_USES_PYTHON` was not enabled on Windows. The problem (that should have been fixed in https://github.com/robotology/robotology-superbuild/pull/663 but it wasn't) is that in https://github.com/robotology/robotology-superbuild/blob/5992ad105a156571461426525aecea3361c434dd/.ci/initial-cache.gh.cmake#L7 we always set `ROBOTOLOGY_USES_PYTHON` to OFF on Windows. This was fixed in https://github.com/robotology/robotology-superbuild/pull/749/commits/86a45c78e5869b219fb538aed48e8e1cbc30988b , and the related Generated Conda Recipes job seems to be finally ok: https://github.com/robotology/robotology-superbuild/actions/runs/867148763 .

**Comment by traversaro** [comment]
Finally successful: https://github.com/robotology/robotology-superbuild/runs/2650809874 .

**Comment by diegoferigo** [comment]
Awesome!! Thanks for finalizing this much awaited feature :rocket:

**Comment by diegoferigo** [comment]
I guess that for scenario / gym-ignition, the most simple way to bring them to conda is directly using conda-forge. I believe that the complexity of handling and maintaining the Ignition dependency inside the superbuild is too much.

**Comment by traversaro** [comment]
> 
> 
> I guess that for scenario / gym-ignition, the most simple way to bring them to conda is directly using conda-forge. I believe that the complexity of handling and maintaining the Ignition dependency inside the superbuild is too much.

Yes, probably that is the case, see also https://github.com/robotology/robotology-superbuild/issues/752 .

**Pull request merged**